config/configtls/configtls_test.go

```diff
@@ -201,3 +206,166 @@ func TestLoadTLSServerConfig(t *testing.T) {
 	assert.NoError(t, err)
 	assert.NotNil(t, tlsCfg)
 }
+
+func TestCertificateReload(t *testing.T) {
+	t.Run("Should eagerly load the given certificate", func(t *testing.T) {
+		options := TLSSetting{
+			CertFile: path.Join("testdata", "test-cert.pem"),
+			KeyFile:  path.Join("testdata", "test-key.pem"),
+		}
+		cfg, err := options.loadTLSConfig()
+		assert.NoError(t, err)
+		assert.NotNil(t, cfg)
+		cert, err := cfg.GetCertificate(&tls.ClientHelloInfo{})
+		assert.NoError(t, err)
+		assert.NotNil(t, cert)
+		pCert, err := x509.ParseCertificate(cert.Certificate[0])
+		assert.NoError(t, err)
+		assert.NotNil(t, pCert)
+		assert.Equal(t, "mkcert tristan@tristans", pCert.Issuer.CommonName)
+	})
+
+	t.Run("Should reload the certificate after the reload-interval", func(t *testing.T) {
+		// Copy certs into a temp dir so we can safely modify them
+		certFile, err := os.CreateTemp("", "cert")
+		assert.NoError(t, err)
+		defer os.Remove(certFile.Name())
+
+		keyFile, err := os.CreateTemp("", "key")
+		assert.NoError(t, err)
+		defer os.Remove(keyFile.Name())
+
+		fdc, err := os.Open(path.Join("testdata", "test-cert.pem"))
+		assert.NoError(t, err)
+		_, err = io.Copy(certFile, fdc)
+		assert.NoError(t, err)
+
+		fdk, err := os.Open(path.Join("testdata", "test-key.pem"))
+		assert.NoError(t, err)
+		_, err = io.Copy(keyFile, fdk)
+		assert.NoError(t, err)
+		assert.NoError(t, fdk.Close())
+
+		options := TLSSetting{
+			CertFile:       certFile.Name(),
+			KeyFile:        keyFile.Name(),
+			ReloadInterval: 100 * time.Millisecond,
+		}
+		cfg, err := options.loadTLSConfig()
+		assert.NoError(t, err)
+		assert.NotNil(t, cfg)
+
+		// Asssert that we loaded the original certificate
+		cert, err := cfg.GetCertificate(&tls.ClientHelloInfo{})
+		assert.NoError(t, err)
+		assert.NotNil(t, cert)
+		pCert, err := x509.ParseCertificate(cert.Certificate[0])
+		assert.NoError(t, err)
+		assert.NotNil(t, pCert)
+		assert.Equal(t, "mkcert tristan@tristans", pCert.Issuer.CommonName)
+
+		// Change the certificate
+		assert.NoError(t, certFile.Truncate(0))
+		assert.NoError(t, keyFile.Truncate(0))
+		_, err = certFile.Seek(0, 0)
+		assert.NoError(t, err)
+		_, err = keyFile.Seek(0, 0)
+		assert.NoError(t, err)
+
+		fdc2, err := os.Open(path.Join("testdata", "test-cert2.pem"))
+		assert.NoError(t, err)
+		_, err = io.Copy(certFile, fdc2)
+		assert.NoError(t, err)
+		assert.NoError(t, fdc2.Close())
+
+		fdk2, err := os.Open(path.Join("testdata", "test-key2.pem"))
+		assert.NoError(t, err)
+		_, err = io.Copy(keyFile, fdk2)
+		assert.NoError(t, err)
+		assert.NoError(t, fdk2.Close())
+
+		// Wait ReloadInterval to ensure a reload will happen
+		time.Sleep(options.ReloadInterval)
```

Comment:
This might make this test unreliable, as it might take longer to read/reload than the actual reload interval. How about allowing consumers to be called back whenever the cert is reloaded? You can then register a callback here in the test, and use it with a wait group instead of relying on unreliable timings.

Reply:
If you look at the code this tests, you'll see that this is an entirely synchronous function.
`GetCertificates` reloads the certificates lazily, the operation is only performed if a new certificate is requested. The `ReloadInterval` config parameter is there to ensure we only reload once in a while rather than every time a cert is needed, as this will impact performance.

By sleeping for `ReloadInterval` here we ensure the time since the last reload is strictly larger than `ReloadInterval` (since we do some other work before) and the cert will always be reloaded.
This test will never fail, the sleep is not there to pave over a potential async race condition.

Reply:
So, your test is waiting 100ms for nothing? That's not nice to other tests ;-) Wouldn't the same work if you set it to, say, 100 Microseconds? Or, anything higher than 0, for that matter...

Reply:
Any value will do, it can definitely be shorter.